stream_alert/rule_processor/payload.py

```diff
@@ -0,0 +1,310 @@
+                    'record:{}>').format(self.__class__.__name__, self.valid,
+                                         self.log_source, self.entity,
+                                         self.type, self.records)
+
+        return repr_str
+
+    @abstractproperty
+    def service(self):
+        """Read only service property enforced on subclasses.
+
+        Returns:
+            [string] The service name for this payload type.
+        """
+
+    @abstractmethod
+    def pre_parse(self):
+        """Pre-parsing method that should be implemented by all subclasses.
+        This establishes the `pre_parsed_record` property to allow for parsing.
+
+        Returns:
+            [generator] Yields instances of `self` back to the caller with the
+                proper `pre_parsed_record` set. Conforming to the interface of
+                returning a generator provides the ability to support multi-record
+                payloads, such as those similar to S3.
+        """
+
+    def _refresh_record(self, new_record):
+        """Replace the currently loaded record with a new one.
+
+        Used mainly when S3 is used as a source, due to looping over files
+        downloadd from S3 events versus all records being readily available
+        from a Kinesis stream.
+
+        Args:
+            new_record [string]: A new raw record to be parsed
+        """
+        self.pre_parsed_record = new_record
+        self.log_source = None
+        self.records = None
+        self.type = None
+        self.valid = False
+
+class S3ObjectSizeError(Exception):
+    """Exception indicating the S3 object is too large to process"""
+
+
+class S3Payload(StreamPayload):
+    """S3Payload class"""
+    s3_object_size = 0
+
+    def service(self): return 's3'
+
+    def pre_parse(self):
+        """Pre-parsing method for S3 objects that will download the s3 object,
+        open it for reading and iterate over lines (records) in the file.
+        This yields back references of this S3Payload instance to the caller
+        with a propertly set `pre_parsed_record` for this record.
+
+        Returns:
+            [generator] Yields instances of `self` back to the caller with the
+                proper `pre_parsed_record` set. Conforms to the interface of
+                returning a generator, providing the ability to support
+                multi-record like this (s3).
+        """
+        s3_file = self._get_object()
+        line_num, processed_size = 0, 0
+        for line_num, data in self._read_downloaded_s3_object(s3_file):
+
+            self._refresh_record(data)
+            yield self
+
+            # Only do the extra calculations below if debug logging is enabled
+            if not LOGGER.isEnabledFor(log_level_debug):
+                continue
+
+            # Add the current data to the total processed size
+            # +1 to account for line feed
+            processed_size += (len(data) + 1)
```

Comment:
@mime-frame `len()` would, if the LF were there.. the LF is not returned as part of the line, since the func `_read_downloaded_s3_object` does a `line.rstrip()` on the line. We could return the line, get the size, then call `rstrip()` on it, but it felt safer to keep string manipulation within the function doing the reading.